src/Http/ServerRequest.php

```diff
@@ -577,16 +584,59 @@ public function session(Session $session = null)
      */
     public function clientIp()
     {
-        if ($this->trustProxy && $this->getEnv('HTTP_X_FORWARDED_FOR')) {
-            $addresses = explode(',', $this->getEnv('HTTP_X_FORWARDED_FOR'));
-            $ipaddr = end($addresses);
-        } elseif ($this->trustProxy && $this->getEnv('HTTP_CLIENT_IP')) {
-            $ipaddr = $this->getEnv('HTTP_CLIENT_IP');
-        } else {
-            $ipaddr = $this->getEnv('REMOTE_ADDR');
+        if ($this->trustProxy) {
+            $forwarded = $this->getEnv('HTTP_X_FORWARDED_FOR');
+            if ($forwarded) {
+                $addresses = array_map('trim', explode(',', $forwarded));
+                $trusted = (count($this->trustedProxies) > 0);
+                $n = count($addresses);
+
+                if ($trusted) {
+                    for ($i = 1; $i < $n; $i++) {
+                        if (in_array($addresses[$i], $this->trustedProxies) === false) {
+                            $trusted = false;
+                            break;
+                        }
+                    }
+                }
+
+                if ($trusted) {
+                    return $addresses[0];
+                }
+
+                return $addresses[$n - 1];
+            }
+
+            if ($this->getEnv('HTTP_X_REAL_IP')) {
+                return $this->getEnv('HTTP_X_REAL_IP');
```

Comment:
By returning early you changed all return signatures here. Note the missing string casts here.